File: clients/android/android/app/src/main/java/app/remotedesktop/client/host/NameConstraints.kt
package app.remotedesktop.client.host

import java.security.cert.X509Certificate
import java.util.Locale

/**
 * Wofür die eigene Stelle Zertifikate unterschreiben darf — X.509 Name
 * Constraints (RFC 5280, 4.2.1.10), kritisch.
 *
 * **Der Befund dahinter (Durchsicht B5):** die CA dieses Handys landet am
 * Rechner in der Liste vertrauter Stellen und am anderen Handy im
 * Zertifikatspeicher des Benutzers, dem auch Chrome vertraut. Ohne
 * Einschränkung könnte, wer den Schlüssel dieser Stelle hat, ein Zertifikat
 * für `google.de` unterschreiben. Mit den Einschränkungen hier gilt sie nur
 * für private Adressen, die üblichen Heimnetz-Endungen, Tailscale und die
 * eigenen Namen. Dasselbe tut der Rechner in `agent/Services/NameConstraints.cs`.
 */
object NameConstraints {

    const val OID = "2.5.29.30"

    /** Endungen, unter denen ein Gerät zu Hause heißt — ohne Punkt davor. */
    val DEFAULT_SUFFIXES = listOf(
        "localhost", "local", "lan", "home", "internal", "fritz.box", "home.arpa", "ts.net",
    )

    /** Private Bereiche, Loopback und Link-Local, als Adresse und Präfixlänge. */
    val DEFAULT_RANGES: List<Pair<ByteArray, Int>> = listOf(
        byteArrayOf(10, 0, 0, 0) to 8,
        byteArrayOf(172.toByte(), 16, 0, 0) to 12,
        byteArrayOf(192.toByte(), 168.toByte(), 0, 0) to 16,
        byteArrayOf(100, 64, 0, 0) to 10,
        byteArrayOf(169.toByte(), 254.toByte(), 0, 0) to 16,
        byteArrayOf(127, 0, 0, 0) to 8,
    )

    private const val TAG_DNS = 0x82
    private const val TAG_IP = 0x87

    /**
     * Der Wert der Erweiterung für eine neue Stelle: die Vorgaben plus alles,
     * was [names] darüber hinaus nennt.
     */
    fun build(names: List<String>): ByteArray {
        val dns = DEFAULT_SUFFIXES.toMutableList()
        val ranges = DEFAULT_RANGES.toMutableList()

        for (raw in names) {
            val name = normalize(raw)

            if (name.isEmpty()) {
                continue
            }

            val address = parseIpv4(name)

            if (address != null) {
                if (ranges.none { (network, prefix) -> covers(network, prefix, address) }) {
                    ranges += address to 32
                }
            } else if (dns.none { matchesSuffix(name, it) }) {
                dns += name
            }
        }

        val subtrees = dns.map { Der.sequence(Der.implicit(2, it.toByteArray(Charsets.US_ASCII))) } +
            ranges.map { (network, prefix) -> Der.sequence(Der.implicit(7, network + mask(prefix))) }

        // permittedSubtrees [0] IMPLICIT SEQUENCE OF GeneralSubtree
        return Der.sequence(Der.implicitSequence(0, subtrees.fold(ByteArray(0)) { acc, part -> acc + part }))
    }

    /**
     * Ob die Stelle für diese Namen unterschreiben darf. Eine Stelle ohne die
     * Erweiterung — von vor v1.4 — darf alles; sie bleibt stehen, damit
     * bestehende Kopplungen gültig bleiben.
     */
    fun permits(authority: X509Certificate, names: List<String>): Boolean {
        val raw = authority.getExtensionValue(OID) ?: return true
        val (dns, ranges) = read(raw) ?: return false

        return names.map(::normalize).filter { it.isNotEmpty() }.all { name ->
            val address = parseIpv4(name)

            if (address != null) {
                ranges.any { (network, prefix) -> covers(network, prefix, address) }
            } else {
                dns.any { matchesSuffix(name, it) }
            }
        }
    }

    /** Liest die erlaubten Namen und Bereiche aus dem Wert der Erweiterung. */
    private fun read(raw: ByteArray): Pair<List<String>, List<Pair<ByteArray, Int>>>? = runCatching {
        val dns = mutableListOf<String>()
        val ranges = mutableListOf<Pair<ByteArray, Int>>()

        // OCTET STRING → SEQUENCE → [0] → SEQUENCE OF GeneralSubtree
        val outer = element(raw, 0)
        val constraints = element(outer.content, 0)
        var offset = 0

        while (offset < constraints.content.size) {
            val part = element(constraints.content, offset)
            offset = part.next

            if (part.tag != 0xA0) {
                continue
            }

            var inner = 0

            while (inner < part.content.size) {
                val subtree = element(part.content, inner)
                inner = subtree.next

                val base = element(subtree.content, 0)

                when (base.tag) {
                    TAG_DNS -> dns += normalize(String(base.content, Charsets.US_ASCII))
                    TAG_IP -> if (base.content.size == 8) {
                        ranges += base.content.copyOfRange(0, 4) to prefixOf(base.content.copyOfRange(4, 8))
                    }
                }
            }
        }

        dns.toList() to ranges.toList()
    }.getOrNull()

    private class Element(val tag: Int, val content: ByteArray, val next: Int)

    /** Ein DER-Element ab [offset]: Kennung, Inhalt, und wo das nächste beginnt. */
    private fun element(bytes: ByteArray, offset: Int): Element {
        val tag = bytes[offset].toInt() and 0xFF
        var position = offset + 1
        var length = bytes[position].toInt() and 0xFF
        position++

        if (length and 0x80 != 0) {
            val count = length and 0x7F
            length = 0

            repeat(count) {
                length = (length shl 8) or (bytes[position].toInt() and 0xFF)
                position++
            }
        }

        return Element(tag, bytes.copyOfRange(position, position + length), position + length)
    }

    private fun matchesSuffix(name: String, suffix: String): Boolean =
        name == suffix || name.endsWith(".$suffix")

    private fun covers(network: ByteArray, prefix: Int, address: ByteArray): Boolean {
        if (network.size != address.size) {
            return false
        }

        for (bit in 0 until prefix) {
            val mask = 0x80 ushr (bit % 8)

            if ((network[bit / 8].toInt() and mask) != (address[bit / 8].toInt() and mask)) {
                return false
            }
        }

        return true
    }

    private fun mask(prefix: Int): ByteArray {
        val mask = ByteArray(4)

        for (bit in 0 until prefix) {
            mask[bit / 8] = (mask[bit / 8].toInt() or (0x80 ushr (bit % 8))).toByte()
        }

        return mask
    }

    private fun prefixOf(mask: ByteArray): Int {
        var prefix = 0

        for (value in mask) {
            for (bit in 7 downTo 0) {
                if ((value.toInt() and (1 shl bit)) == 0) {
                    return prefix
                }

                prefix++
            }
        }

        return prefix
    }

    internal fun parseIpv4(text: String): ByteArray? {
        val parts = text.split('.')

        if (parts.size != 4) {
            return null
        }

        val bytes = ByteArray(4)

        for (index in parts.indices) {
            val value = parts[index].toIntOrNull() ?: return null

            if (value !in 0..255 || (parts[index].length > 1 && parts[index].startsWith("0"))) {
                return null
            }

            bytes[index] = value.toByte()
        }

        return bytes
    }

    private fun normalize(name: String): String = name.trim().trim('[', ']').lowercase(Locale.ROOT)
}

File: agent/Services/NameConstraints.cs
using System.Formats.Asn1;
using System.Net;
using System.Net.Sockets;
using System.Security.Cryptography;
using System.Security.Cryptography.X509Certificates;

namespace RemoteDesktopAgent.Services;

/// <summary>
/// Wofür die eigene Stelle Zertifikate unterschreiben darf — X.509 Name
/// Constraints (RFC 5280, 4.2.1.10), kritisch.
///
/// <para>
/// **Der Befund dahinter (Durchsicht B5):** die CA eines Rechners landet am
/// Handy im Zertifikatspeicher des Benutzers, dem auch Chrome vertraut. Ohne
/// Einschränkung könnte, wer den Schlüssel dieser Stelle hat, ein Zertifikat
/// für <c>google.de</c> unterschreiben, und das Handy nähme es an. Mit den
/// Einschränkungen hier gilt die Stelle nur für private Adressen, die
/// üblichen Heimnetz-Endungen, Tailscale und den eingetragenen Namen —
/// alles andere lehnt jeder Client ab, der die Erweiterung versteht, und
/// weil sie kritisch ist, auch jeder, der sie nicht versteht.
/// </para>
/// </summary>
public static class NameConstraints
{
    public static readonly Oid Id = new("2.5.29.30");

    /// <summary>
    /// Endungen, unter denen ein Rechner zu Hause heißt. Ohne Punkt davor: eine
    /// Einschränkung „fritz.box" erlaubt <c>pc.fritz.box</c> und <c>fritz.box</c>.
    /// </summary>
    public static readonly IReadOnlyList<string> DefaultSuffixes =
        ["localhost", "local", "lan", "home", "internal", "fritz.box", "home.arpa", "ts.net"];

    /// <summary>Private Adressbereiche, Loopback und Link-Local — v4 und v6.</summary>
    public static readonly IReadOnlyList<(IPAddress Network, int Prefix)> DefaultRanges =
    [
        (IPAddress.Parse("10.0.0.0"), 8),
        (IPAddress.Parse("172.16.0.0"), 12),
        (IPAddress.Parse("192.168.0.0"), 16),
        (IPAddress.Parse("100.64.0.0"), 10),
        (IPAddress.Parse("169.254.0.0"), 16),
        (IPAddress.Parse("127.0.0.0"), 8),
        (IPAddress.Parse("::1"), 128),
        (IPAddress.Parse("fc00::"), 7),
        (IPAddress.Parse("fe80::"), 10)
    ];

    private static readonly Asn1Tag Permitted = new(TagClass.ContextSpecific, 0, isConstructed: true);
    private static readonly Asn1Tag DnsName = new(TagClass.ContextSpecific, 2);
    private static readonly Asn1Tag IpAddress = new(TagClass.ContextSpecific, 7);

    /// <summary>
    /// Die Erweiterung für eine neue Stelle: die Vorgaben plus alles, was
    /// <paramref name="names"/> darüber hinaus nennt — ein Name als Name, eine
    /// Adresse außerhalb der Bereiche als einzelne Adresse.
    /// </summary>
    public static X509Extension Build(IEnumerable<string> names)
    {
        var dns = new List<string>(DefaultSuffixes);
        var ranges = new List<(IPAddress Network, int Prefix)>(DefaultRanges);

        foreach (var raw in names)
        {
            var name = Normalize(raw);

            if (name.Length == 0)
            {
                continue;
            }

            if (IPAddress.TryParse(name, out var address))
            {
                if (!ranges.Any(range => Covers(range.Network, range.Prefix, address)))
                {
                    ranges.Add((address, address.AddressFamily == AddressFamily.InterNetwork ? 32 : 128));
                }
            }
            else if (!dns.Any(suffix => MatchesSuffix(name, suffix)))
            {
                dns.Add(name);
            }
        }

        var writer = new AsnWriter(AsnEncodingRules.DER);

        using (writer.PushSequence())
        using (writer.PushSequence(Permitted))
        {
            foreach (var name in dns)
            {
                using (writer.PushSequence())
                {
                    writer.WriteCharacterString(UniversalTagNumber.IA5String, name, DnsName);
                }
            }

            foreach (var (network, prefix) in ranges)
            {
                using (writer.PushSequence())
                {
                    writer.WriteOctetString(WithMask(network, prefix), IpAddress);
                }
            }
        }

        return new X509Extension(Id, writer.Encode(), critical: true);
    }

    /// <summary>
    /// Ob die Stelle Zertifikate für diese Namen unterschreiben darf.
    ///
    /// Eine Stelle ohne die Erweiterung — von vor v1.4 — darf alles; sie
    /// bleibt stehen, damit bestehende Kopplungen gültig bleiben. Erst wer neu
    /// koppelt, bekommt eine eingeschränkte.
    /// </summary>
    public static bool Permits(X509Certificate2 authority, IEnumerable<string> names)
    {
        var extension = authority.Extensions[Id.Value!];

        if (extension is null)
        {
            return true;
        }

        var (dns, ranges) = Read(extension.RawData);

        foreach (var raw in names)
        {
            var name = Normalize(raw);

            if (name.Length == 0)
            {
                continue;
            }

            var allowed = IPAddress.TryParse(name, out var address)
                ? ranges.Any(range => Covers(range.Network, range.Prefix, address))
                : dns.Any(suffix => MatchesSuffix(name, suffix));

            if (!allowed)
            {
                return false;
            }
        }

        return true;
    }

    private static (List<string> Dns, List<(IPAddress Network, int Prefix)> Ranges) Read(byte[] raw)
    {
        var dns = new List<string>();
        var ranges = new List<(IPAddress, int)>();

        try
        {
            var reader = new AsnReader(raw, AsnEncodingRules.DER).ReadSequence();

            if (!reader.HasData || !reader.PeekTag().HasSameClassAndValue(Permitted))
            {
                return (dns, ranges);
            }

            var subtrees = reader.ReadSequence(Permitted);

            while (subtrees.HasData)
            {
                var subtree = subtrees.ReadSequence();
                var tag = subtree.PeekTag();

                if (tag.HasSameClassAndValue(DnsName))
                {
                    dns.Add(Normalize(subtree.ReadCharacterString(UniversalTagNumber.IA5String, DnsName)));
                }
                else if (tag.HasSameClassAndValue(IpAddress))
                {
                    var bytes = subtree.ReadOctetString(IpAddress);
                    var half = bytes.Length / 2;

                    if (half is 4 or 16)
                    {
                        ranges.Add((new IPAddress(bytes[..half]), PrefixOf(bytes[half..])));
                    }
                }
            }
        }
        catch (AsnContentException)
        {
            // Eine unlesbare Einschränkung erlaubt nichts — dann wird die
            // Stelle neu ausgestellt, sichtbar für die Clients.
        }

        return (dns, ranges);
    }

    private static bool MatchesSuffix(string name, string suffix) =>
        name.Equals(suffix, StringComparison.Ordinal)
        || name.EndsWith("." + suffix, StringComparison.Ordinal);

    private static bool Covers(IPAddress network, int prefix, IPAddress address)
    {
        if (network.AddressFamily != address.AddressFamily)
        {
            return false;
        }

        var a = network.GetAddressBytes();
        var b = address.GetAddressBytes();

        for (var bit = 0; bit < prefix; bit++)
        {
            var mask = (byte)(0x80 >> (bit % 8));

            if ((a[bit / 8] & mask) != (b[bit / 8] & mask))
            {
                return false;
            }
        }

        return true;
    }

    private static byte[] WithMask(IPAddress network, int prefix)
    {
        var address = network.GetAddressBytes();
        var mask = new byte[address.Length];

        for (var bit = 0; bit < prefix; bit++)
        {
            mask[bit / 8] |= (byte)(0x80 >> (bit % 8));
        }

        return [.. address, .. mask];
    }

    private static int PrefixOf(byte[] mask)
    {
        var prefix = 0;

        foreach (var value in mask)
        {
            for (var bit = 7; bit >= 0; bit--)
            {
                if ((value & (1 << bit)) == 0)
                {
                    return prefix;
                }

                prefix++;
            }
        }

        return prefix;
    }

    private static string Normalize(string name) => name.Trim().Trim('[', ']').ToLowerInvariant();
}
